lib/pbench/server/database/models/datasets.py

```diff
@@ -671,48 +707,167 @@ class Metadata(Database.Base):
+    # static methods (used consistently by the API layer) allow client code
+    # to interact with these either as a complete JSON document ("user") or
+    # as a full dotted key path ("user.contact.name.first") or at any JSON
+    # layer between.
+    #
+    # API keyword validation uses the trailing ".*" here to indicate that only
+    # the first element of the path should be validated, allowing the client
+    # a completely uninterpreted namespace below that.
+    #
+    # {"user.cloud": "AWS", "user.mood": "CLOUDY"}}
+    USER = "user.*"
+
     # REINDEX boolean flag to indicate when a dataset should be re-indexed
     #
-    # {"REINDEX": True}
-    REINDEX = "REINDEX"
+    # {"server.reindex": True}
+    REINDEX = "server.reindex"
 
     # ARCHIVED boolean flag to indicate when a tarball has been archived
     #
-    # {"ARCHIVED": True}
-    ARCHIVED = "ARCHIVED"
+    # {"server.archived": True}
+    ARCHIVED = "server.archived"
 
     # TARBALL_PATH access path of the dataset tarball. (E.g., we could use this
     # to record an S3 object store key.) NOT YET USED.
     #
-    # {"TARBALL_PATH": "/srv/pbench/archive/fs-version-001/ctrl/example__2021.05.21T07.15.27.tar.xz"}
-    TARBALL_PATH = "TARBALL_PATH"
-
-    # INDEX_MAP a two-level dict recording the MD5 document IDs for each
-    # Elasticsearch index containing documents for this dataset.
+    # {
+    #   "server.tarball-path": "/srv/pbench/archive/fs-version-001/"
+    #       "ctrl/example__2021.05.21T07.15.27.tar.xz"
+    # }
+    TARBALL_PATH = "server.tarball-path"
+
+    # INDEX_MAP a dict recording the set of MD5 document IDs for each
+    # Elasticsearch index that contains documents for this dataset.
     #
-    # {"drb.v6.run-data.2021-07": ["MD5"], "drb.v6.run-toc.2021-07": ["MD5-1", "MD5-2"]}
-    INDEX_MAP = "INDEX_MAP"
+    # {
+    #    "server.index-map": {
+    #      "drb.v6.run-data.2021-07": ["MD5"],
+    #      "drb.v6.run-toc.2021-07": ["MD5-1", "MD5-2"]
+    #    }
+    # }
+    INDEX_MAP = "server.index-map"
+
+    # --- Standard Metadata keys
+
+    # Metadata keys that clients can update
+    USER_UPDATEABLE_METADATA = [DASHBOARD, USER]
 
-    METADATA_KEYS = [REINDEX, ARCHIVED, TARBALL_PATH, INDEX_MAP]
+    # Metadata keys that are accessible to clients
+    USER_METADATA = USER_UPDATEABLE_METADATA + [DELETION]
+
+    # Metadata keys that are for internal use only
+    INTERNAL_METADATA = [REINDEX, ARCHIVED, TARBALL_PATH, INDEX_MAP]
+
+    # All supported Metadata keys
+    METADATA_KEYS = USER_METADATA + INTERNAL_METADATA
+
+    # NOTE: the ECMA JSON specification allows JSON "names" (keys) to be any
+    # string, though most implementations and schemas enforce or recommend
+    # stricter syntax. Pbench restricts dataset metadata open namespace keys to
+    # lowercase letters, numbers, underscore, and hyphen, with keys separated
+    # by a period.
+    _valid_key_charset = re.compile(r"[a-z0-9_.-]+")
```

Comment:
If you find some other reason to make a further change, consider upping this RE to something like this:
```
    _valid_key_charset = re.compile(r"[a-z0-9_-]+(?:\.[a-z0-9_-]+)*")
```
That is, a sequence of one or more of our favorite (non-dot) characters followed by zero or more such sequences, with each sequence separated by dots.

Reply:
I thought about that for about 10 seconds when I wrote it, and decided that the extra complication was pointless unless we're going to impose some particular syntax on the keys -- e.g., must lead with a letter, no leading or trailing `_` or `-`, stuff like that. And I considered that, too, and decided it wasn't that important and I just didn't want to bother.

True, on the other hand, it's true that I wouldn't have to look for empty keys in the split path because the RE would prevent that. I'll consider this, too, in the other PR, but I'd rather get this one in than mess with it more.

Reply:
Yeah, it was excluding the empty keys (including leading and trailing dots) which made it appeal, but you can change it in the next PR.